pkg/tuned/controller.go

```diff
@@ -384,53 +389,68 @@ func profilesSync(profiles []tunedv1.TunedProfile, recommendedProfile string) (b
-	for profile := range recommendedProfileDeps {
+	dirEntries, err := os.ReadDir(tunedProfilesDirCustom)
+	if err != nil {
+		return change, err
+	}
+
+	// Deal with TuneD profiles absent from Tuned CRs, but still present in <tunedProfilesDirCustom>/<profile>/
+	for _, dirEntry := range dirEntries {
+		profile := dirEntry.Name()
+		if !dirEntry.IsDir() {
+			// There shouldn't be anything but directories in <tunedProfilesDirCustom>, but if there is, skip it.
+			continue
+		}
+
 		if len(profile) == 0 {
+			// This should never happen, but if it does, do not wipe the entire tunedProfilesDirCustom directory.
 			continue
 		}
 
-		if !extractedNew[profile] {
-			// TuneD profile does not exist in the Tuned CR, but the recommended profile depends on it.
-			profileDir := fmt.Sprintf("%s/%s", tunedProfilesDirCustom, profile)
-			if _, err := os.Stat(profileDir); err == nil {
-				// We have a stale TuneD profile directory in /etc/tuned/<profile>/
-				// Remove it.
-				err := os.RemoveAll(profileDir)
-				if err != nil {
-					return change, fmt.Errorf("failed to remove %q: %v", profileDir, err)
-				}
-				change = true
-				klog.Infof("removed TuneD profile %q", profileDir)
-			}
+		if extractedNew[profile] {
+			// Do not delete the freshly extracted profiles.  These are the profiles present in the Profile CR.
+			continue
+		}
+		// This TuneD profile does not exist in the Profile CR, delete it.
+		profileDir := fmt.Sprintf("%s/%s", tunedProfilesDirCustom, profile)
+		err := os.RemoveAll(profileDir)
+		if err != nil {
+			return change, fmt.Errorf("failed to remove %q: %v", profileDir, err)
+		}
+		klog.Infof("removed TuneD profile %q", profileDir)
+
+		if recommendedProfileDeps[profile] {
+			// This TuneD profile does not exist in the Profile CR, but the recommended profile depends on it.
+			// Trigger a change to report a configuration issue -- we depend on a profile that does not exist.
+			change = true
 		}
 	}
 
 	return change, nil
 }
 
-// providerExtract extracts Cloud Provider name into openshiftTunedProvider file.
+// providerExtract extracts Cloud Provider name into ocpTunedProvider file.
 func providerExtract(provider string) error {
-	klog.Infof("extracting cloud provider name to %v", openshiftTunedProvider)
+	klog.Infof("extracting cloud provider name to %v", ocpTunedProvider)
 
-	f, err := os.Create(openshiftTunedProvider)
+	f, err := os.Create(ocpTunedProvider)
 	if err != nil {
-		return fmt.Errorf("failed to create cloud provider name file %q: %v", openshiftTunedProvider, err)
+		return fmt.Errorf("failed to create cloud provider name file %q: %v", ocpTunedProvider, err)
 	}
 	defer f.Close()
 	if _, err = f.WriteString(provider); err != nil {
-		return fmt.Errorf("failed to write cloud provider name file %q: %v", openshiftTunedProvider, err)
+		return fmt.Errorf("failed to write cloud provider name file %q: %v", ocpTunedProvider, err)
 	}
 
 	return nil
```

Comment:
non-blocking:  this seems another case of function which can be almost completely collapsed into a `os.WriteFile` call.

Reply:
Yes, I know. I didn't want to change it yesterday because I believe that should be part of a cleanup and not this PR. Adding on my TODO list.